okhttp/src/test/java/okhttp3/internal/http2/HttpOverHttp2Test.java

```diff
@@ -1789,30 +1793,46 @@ public void connectionShutdownAfterHealthCheck(
   public void responseHeadersAfterGoaway(
       Protocol protocol, MockWebServer mockWebServer) throws Exception {
     setUp(protocol, mockWebServer);
-    // Flaky https://github.com/square/okhttp/issues/4836
     server.enqueue(new MockResponse()
-        .setHeadersDelay(1, SECONDS)
-        .setBody("ABC"));
+      .setHeadersDelay(1, SECONDS)
+      .setBody("ABC"));
     server.enqueue(new MockResponse()
-        .setSocketPolicy(SocketPolicy.DISCONNECT_AT_END)
-        .setBody("DEF"));
+      .setSocketPolicy(SocketPolicy.DISCONNECT_AT_END)
+      .setBody("DEF"));
+
+    CountDownLatch latch = new CountDownLatch(2);
+    ArrayList<IOException> errors = new ArrayList<>();
 
     BlockingQueue<String> bodies = new LinkedBlockingQueue<>();
     Callback callback = new Callback() {
       @Override public void onResponse(Call call, Response response) throws IOException {
         bodies.add(response.body().string());
+        latch.countDown();
       }
 
       @Override public void onFailure(Call call, IOException e) {
-        System.out.println(e);
+        errors.add(e);
+        latch.countDown();
       }
     };
     client.newCall(new Request.Builder().url(server.url("/")).build()).enqueue(callback);
     client.newCall(new Request.Builder().url(server.url("/")).build()).enqueue(callback);
 
-    assertThat(bodies.poll(3, SECONDS)).isEqualTo("DEF");
-    assertThat(bodies.poll(3, SECONDS)).isEqualTo("ABC");
-    assertThat(server.getRequestCount()).isEqualTo(2);
+    latch.await();
+
+    assertThat(bodies.remove()).isEqualTo("DEF");
+
+    if (errors.isEmpty()) {
```

Comment:
I don't know the root cause for the flakiness in this case, but I suggest that you look at a few techniques I used in the past to make similar tests deterministic and that may apply here: https://github.com/square/okhttp/pull/4792, https://github.com/square/okhttp/pull/4728, https://github.com/square/okhttp/pull/4772.

Reply:
Thanks, will do. I wanted to get something in so we have PR and master with working builds.  So I hope you forgive my hackish attempt to start!

Reply:
I will get familiar with QueueDispatcher, there are a couple more flaky tests here.